### PR TITLE
Tfm integration

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Add TFM support in NCS.

This brings TFM into partition manager, though it's partitions cannot be modified, since TFM doesn't yet allow partition sizes/addresses to be set from the outside.

Rebased on #3357 and #3268

### CODE REVIEW

File: samples/tfm_hello_world/README.rst
Comment (b-gent) [inline samples/tfm_hello_world/README.rst:9]: Text to be enhanced by @ioannisg

File: samples/tfm_hello_world/README.rst
Comment (b-gent) [inline samples/tfm_hello_world/README.rst:12]: Described in Zephyr doc https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/tfm_integration/tfm_ipc#on-nrf5340-and-nrf9160

File: samples/tfm_hello_world/README.rst
Comment (b-gent) [inline samples/tfm_hello_world/README.rst:23]:
```suggestion
    Hello World! nrf5340pdk_nrf5340_cpuappns
```

**Comment by b-gent** [comment]
"When built for the nrf5340pdk_nrf5340_cpuappns board, the Secure Partition Manager sample is automatically included in the build."
@oyvindronningstad  is this sentence still true? This comes from: 
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_nrf5340.html#application-core

File: subsys/tfm/Kconfig
Comment (hakonfam) [inline subsys/tfm/Kconfig:7]: Could this be an imply in the original definition of PM_SINGLE_IMAGE?


```
config PM_SINGLE_IMAGE
    imply if BUILD_WITH_TFM
```

File: subsys/partition_manager/pm.yml.tfm
Comment (hakonfam) [inline subsys/partition_manager/pm.yml.tfm:6]: Don't you need to split the TFM ram from the non secure ram?

File: subsys/tfm/CMakeLists.txt
Comment (hakonfam) [inline subsys/tfm/CMakeLists.txt:7]: This looks quite cryptic on its own, could you add a comment explaining why this is needen?

Also, could you double check that we don't have any ZEPHYR_BINARY_DIR or similar aliases for CMAKE_BINARY_DIR that we should use instead? I always mix these up and don't remember when we should use what, but just double check what the idiomatic way is.

File: subsys/partition_manager/pm.yml.tfm
Comment (hakonfam) [reply subsys/partition_manager/pm.yml.tfm:6]: I assume that the limits are hard coded by TFM anyway, and not handled like it is in SPM.

That difference might deserve a comment if people are wondering why SPM use a separate region while TFM does not.

File: CODEOWNERS
Comment (hakonfam) [inline CODEOWNERS:92]: Should this be in `samples/subsys/tfm/tfm_hello_world` ?

File: samples/tfm_hello_world/README.rst
Comment (hakonfam) [inline samples/tfm_hello_world/README.rst:15]: To be more future proof, state Trustzone enabled devices, and say "for instance 91 and 53" that way we don't need to update this again.

File: samples/tfm_hello_world/prj.conf
Comment (hakonfam) [inline samples/tfm_hello_world/prj.conf:1]: Missing copyright.

File: samples/tfm_hello_world/sample.yaml
Comment (hakonfam) [inline samples/tfm_hello_world/sample.yaml:1]: Missing copyright.

File: samples/tfm_hello_world/src/main.c
Comment (hakonfam) [inline samples/tfm_hello_world/src/main.c:2]: Should it be like this?

File: CODEOWNERS
Comment (oyvindronningstad) [reply CODEOWNERS:92]: Will move to samples/tfm/tfm_hello_world

File: subsys/partition_manager/pm.yml.tfm
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.tfm:6]: Sure, I'm thinking of removing the prompt from the Kconfigs and explaining in the help text. Or do you think a comment here is necessary instead/also?

File: subsys/tfm/CMakeLists.txt
Comment (oyvindronningstad) [reply subsys/tfm/CMakeLists.txt:7]: I don't think we do, for this case. @tejlmand can comment if I'm wrong.

File: subsys/tfm/Kconfig
Comment (oyvindronningstad) [reply subsys/tfm/Kconfig:7]: Not sure if `imply if` is a thing, but I can add a `default y if BUILD_WITH_TFM`.

Comment (oyvindronningstad) [reply subsys/tfm/Kconfig:7]: I think I'd actually prefer that it stays here, so the TFM-related stuff is mostly in the same place.

Comment (oyvindronningstad) [reply subsys/tfm/Kconfig:7]: Wait, the partition sizes are already in there. Sure, why not.

Comment (oyvindronningstad) [reply subsys/tfm/Kconfig]: I even moved the CMakeLists stuff into the partition_manager subsys, see if it's ok.

File: subsys/partition_manager/CMakeLists.txt
Comment (tejlmand) [inline subsys/partition_manager/CMakeLists.txt:64]: Using `${CMAKE_BINARY_DIR}/tfm/bin` is a bit deep knowledge here.
Could we not get this info by other means ?

**Review by tejlmand — COMMENTED** [review]
Just a minor question.

File: subsys/partition_manager/CMakeLists.txt
Comment (oyvindronningstad) [reply subsys/partition_manager/CMakeLists.txt:64]: I'm experimenting with adding paths as target properties on `tfm`, for the native ns pr. What do you think of that solution?

File: samples/tfm_hello_world/README.rst
Comment (oyvindronningstad) [reply samples/tfm_hello_world/README.rst]: The output actually shows as specified. Not sure why.

File: samples/tfm_hello_world/prj.conf
Comment (oyvindronningstad) [reply samples/tfm_hello_world/prj.conf]: Fixed

File: samples/tfm_hello_world/sample.yaml
Comment (oyvindronningstad) [reply samples/tfm_hello_world/sample.yaml]: Fixed

File: samples/tfm_hello_world/src/main.c
Comment (oyvindronningstad) [reply samples/tfm_hello_world/src/main.c]: Nope :P Fixed

File: subsys/partition_manager/CMakeLists.txt
Comment (oyvindronningstad) [reply subsys/partition_manager/CMakeLists.txt:64]: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/34

Comment (oyvindronningstad) [reply subsys/partition_manager/CMakeLists.txt:64]: I'll leave it as is, since the change won't get downstream in time.

File: samples/tfm_hello_world/README.rst
Comment (oyvindronningstad) [reply samples/tfm_hello_world/README.rst]: It only works on the targets ported to TFM, so only the DKs, not e.g. thingy91.

I could write something like "on TFM-enabled boards like the nRF9160 DK or nRF5340 DK".

File: samples/tfm_hello_world/README.rst
Comment (oyvindronningstad) [reply samples/tfm_hello_world/README.rst]: Building and running will be different from in Zephyr. I added a note about building and running like an ordinary NCS app.

File: samples/tfm_hello_world/README.rst
Comment (oyvindronningstad) [reply samples/tfm_hello_world/README.rst]: Expanded a bit. Take a look.

File: subsys/partition_manager/pm.yml.tfm
Comment (oyvindronningstad) [reply subsys/partition_manager/pm.yml.tfm:6]: Modified the Kconfigs, take a look.

File: samples/tfm_hello_world/README.rst
Comment (oyvindronningstad) [reply samples/tfm_hello_world/README.rst]: Done.

File: include/flash_map_pm.h
Comment (kyberdin) [inline include/flash_map_pm.h:44]: @hakonfam Didn't we talk about this on a different branch yesterday? Was it the same solution?

Comment (oyvindronningstad) [reply include/flash_map_pm.h:44]: Note that this and a bunch of other changes come from #3357 which this is based on. Look at the last commits to see the changes that properly belong to the PR.

**Comment by b-gent** [comment]
I added a reviewed readme.rst file in a separate commit, please squash it before merging.

Approving now, no reason for documentation to hold this PR. We can extend the sample doc slightly (e.g. by adding references) in PR#3409.

**Review by b-gent — APPROVED** [review]

**Review by hakonfam — APPROVED** [review]

File: samples/tfm/tfm_hello_world/README.rst
Comment (kyberdin) [inline samples/tfm/tfm_hello_world/README.rst]: Missing reference? Not sure what the `???` is for here.

Comment (hakonfam) [reply samples/tfm/tfm_hello_world/README.rst]: I think this was fixed in @b-gent 's commit which should be squashed.

File: subsys/partition_manager/CMakeLists.txt
Comment (tejlmand) [reply subsys/partition_manager/CMakeLists.txt:64]: accepted.

File: subsys/partition_manager/Kconfig
Comment (tejlmand) [inline subsys/partition_manager/Kconfig:77]: is this just a `default y` if building with TF-M, or should it actually be enforced the `PM_SINGLE_IMAGE` is selected if `BUILD_WITH_TFM=y` ?

**Review by hurricaneFromMoscow — APPROVED** [review]

File: subsys/partition_manager/Kconfig
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig:77]: Enforced maybe? It doesn't work without.

Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig]: Changed it now, take a look.

File: subsys/partition_manager/Kconfig
Comment (tejlmand) [inline subsys/partition_manager/Kconfig:119]: What if BL2 is is disabled: https://github.com/zephyrproject-rtos/zephyr/blob/master/modules/Kconfig.tfm#L96-L97

Should this setting have a dependency ?

File: subsys/partition_manager/Kconfig
Comment (tejlmand) [reply subsys/partition_manager/Kconfig]: Looks good.

File: subsys/partition_manager/Kconfig
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig:119]: The nRF platforms don't really support that.

**Review by tejlmand — APPROVED** [review]

**Review by kyberdin — APPROVED** [review]

**Comment by oyvindronningstad** [comment]
@thst-nordic Seems we need imgtool here as well: https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/7190/console

**Comment by thst-nordic** [comment]
can you add it here: 
https://github.com/nrfconnect/sdk-nrf/blob/master/scripts/requirements-ci.txt#L27

File: samples/tfm/tfm_hello_world/CMakeLists.txt
Comment (ioannisg) [inline samples/tfm/tfm_hello_world/CMakeLists.txt:7]: @oyvindronningstad doesn't tfm now require cmake > 3.15.0?
does this need to be reflected here? 
@tejlmand ^^

File: samples/tfm/tfm_hello_world/nrf9160dk_nrf9160ns.overlay
Comment (ioannisg) [inline samples/tfm/tfm_hello_world/nrf9160dk_nrf9160ns.overlay:7]: Maybe add a comment why this is needed.

File: samples/tfm/tfm_hello_world/sample.yaml
Comment (ioannisg) [inline samples/tfm/tfm_hello_world/sample.yaml:7]: Do we need _dk_ here instead of _pdk_? or both...

File: samples/tfm/tfm_hello_world/src/main.c
Comment (ioannisg) [inline samples/tfm/tfm_hello_world/src/main.c:16]: guess the #define can start from the beginning of the line

File: subsys/partition_manager/Kconfig
Comment (ioannisg) [reply subsys/partition_manager/Kconfig:119]: Not sure about that - we should be able to build TFM without BL2, in general

**Review by ioannisg — COMMENTED** [review]
Some comments, not really major (worth taking a quick look, if possible)
@oyvindronningstad FYI

File: samples/tfm/tfm_hello_world/CMakeLists.txt
Comment (oyvindronningstad) [reply samples/tfm/tfm_hello_world/CMakeLists.txt:7]: I thought about it. @magnev has also brought this up. I thought maybe that would be covered in the TF-M Cmake files.

File: subsys/partition_manager/Kconfig
Comment (oyvindronningstad) [reply subsys/partition_manager/Kconfig:119]: Maybe, but right now it fails on the signing code in Zephyr.

File: samples/tfm/tfm_hello_world/nrf9160dk_nrf9160ns.overlay
Comment (oyvindronningstad) [reply samples/tfm/tfm_hello_world/nrf9160dk_nrf9160ns.overlay:8]: Thanks, done

File: samples/tfm/tfm_hello_world/sample.yaml
Comment (oyvindronningstad) [reply samples/tfm/tfm_hello_world/sample.yaml]: Thanks, added both.

File: samples/tfm/tfm_hello_world/src/main.c
Comment (oyvindronningstad) [reply samples/tfm/tfm_hello_world/src/main.c]: Done

File: samples/tfm/tfm_hello_world/CMakeLists.txt
Comment (tejlmand) [reply samples/tfm/tfm_hello_world/CMakeLists.txt:7]: yes, it's also covered in TFM CMake file, but when the sample itself is a tfm based sample, then we should inform the user directly, instead of waiting for the TFM cmake part to fail.

Will also work as information to anyone looking directly at the sample/CMakeLists.txt file, that this sample itself requires 3.15.

Things are a bit trickier for samples that can be used both with and without TFM, cause in those cases, 3.15 is the minimum seen from the sample.

Comment (oyvindronningstad) [reply samples/tfm/tfm_hello_world/CMakeLists.txt:7]: I can update it no problem, but I don't want users who enable TFM to have to remember to do this.

I'll update it and add a comment.

Comment (oyvindronningstad) [reply samples/tfm/tfm_hello_world/CMakeLists.txt:7]: Done

File: subsys/partition_manager/Kconfig
Comment (ioannisg) [reply subsys/partition_manager/Kconfig:119]: Ok that's something we need to have documented.
@b-gent FYI

**Review by ioannisg — APPROVED** [review]

**Pull request merged**
